libsql-sqlite3/src/vector.c

```diff
@@ -0,0 +1,558 @@
+**************************************************************************/
+
+size_t vectorDataSize(VectorType type, VectorDims dims){
+  switch( type ){
+    case VECTOR_TYPE_FLOAT32:
+      return dims * sizeof(float);
+    case VECTOR_TYPE_FLOAT64:
+      return dims * sizeof(double);
+    default:
+      assert(0);
+  }
+  return 0;
+}
+
+void vectorInit(Vector *pVector, VectorType type, VectorDims dims, void *data){
+  pVector->type = type;
+  pVector->dims = dims;
+  pVector->data = data;
+  pVector->flags = 0;
+}
+
+/*
+ * Allocate a Vector object and its data buffer
+*/
+Vector *vectorAlloc(VectorType type, VectorDims dims){
+  void *pVector = sqlite3_malloc(sizeof(Vector) + vectorDataSize(type, dims));
+  if( pVector==NULL ){
+    return NULL;
+  }
+  vectorInit(pVector, type, dims, ((char*) pVector) + sizeof(Vector));
+  return pVector;
+}
+
+/*
+ * Allocate a Vector object and its data buffer from the SQLite context. 
+*/
+static Vector* vectorContextAlloc(sqlite3_context *context, int type, int dims){
+  void *pVector = sqlite3_malloc64(sizeof(Vector) + vectorDataSize(type, dims));
+  if( pVector==NULL ){
+    sqlite3_result_error_nomem(context);
+    return NULL;
+  }
+  vectorInit(pVector, type, dims, ((char*) pVector) + sizeof(Vector));
+  return pVector;
+}
+
+/*
+ * Free a Vector object and its data buffer allocated, unless the vector is static.
+*/
+void vectorFree(Vector *pVector){
+  if( pVector==0 ){
+    return;
+  }
+  if( pVector->flags & VECTOR_FLAGS_STATIC ){
+    return;
+  }
+  sqlite3_free(pVector);
+}
+
+float vectorDistanceCos(const Vector *pVector1, const Vector *pVector2){
+  assert( pVector1->type == pVector2->type );
+  switch (pVector1->type) {
+    case VECTOR_TYPE_FLOAT32:
+      return vectorF32DistanceCos(pVector1, pVector2);
+    case VECTOR_TYPE_FLOAT64:
+      return vectorF64DistanceCos(pVector1, pVector2);
+    default:
+      assert(0);
+  }
+  return 0;
+}
+
+/*
+ * Parses vector from text representation (e.g. '[1,2,3]'); vector type must be set
+*/
+static int vectorParseSqliteText(
+  sqlite3_value *arg,
+  Vector *pVector,
+  char **pzErrMsg
+){
+  const unsigned char *pzText;
+  double elem;
+  float *elemsFloat;
+  double *elemsDouble;
+  int iElem = 0;
+  // one more extra character in order to safely print data from elBuf with printf-like method; will be set to zero later
+  char valueBuf[MAX_FLOAT_CHAR_SZ + 1];
+  int iBuf = 0;
+
+  assert( pVector->type == VECTOR_TYPE_FLOAT32 || pVector->type == VECTOR_TYPE_FLOAT64 );
+
+  if( pVector->type == VECTOR_TYPE_FLOAT32 ){
+    elemsFloat = pVector->data;
+  } else if( pVector->type == VECTOR_TYPE_FLOAT64 ){
+    elemsDouble = pVector->data;
+  }
+
+  if( sqlite3_value_type(arg) != SQLITE_TEXT ){
+    *pzErrMsg = sqlite3_mprintf("invalid vector: not a text type");
+    goto error;
+  }
+
+  pzText = sqlite3_value_text(arg);
+  if ( pzText == NULL ) return 0;
+
+  while( pzText && sqlite3Isspace(*pzText) )
```

Comment:
I have some brain problems and can't see why pzText can every be false here.

Reply:
This is not your problem - this is code issue! Will fix this!